### PR TITLE
Add  token create Prehandle 

### PR DESCRIPTION
**Pull request by randered**
**Description**:

Adding PreHandleTokenCreate functionality. 
Created the tests with the old scenarios from SigRequirmentsTests. 

**Related issue(s)**:

Fixes https://github.com/hashgraph/hedera-services/issues/4594
Fixes https://github.com/hashgraph/hedera-services/issues/4377

**Notes for reviewer**:
Added few more test scenarios in order to get Coverage.
Classes for review PreHandleTokenCreateTest & TokenCreateHandler. 
PS. It's My first PR, have mercy 

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by dimitar-dinev — COMMENTED** [review]
Overall LGTM, just a couple of optimizations/nits

File: hedera-node/hedera-token-service-impl/src/test/java/com/hedera/node/app/service/token/impl/test/PreHandleTokenCreateTest.java [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+    }
+
+    private void basicMetaAssertions(
+            final TransactionMetadata meta,
+            final int keysSize,
+            final boolean failed,
+            final ResponseCodeEnum failureStatus) {
+        assertEquals(keysSize, meta.requiredNonPayerKeys().size());
+        assertEquals(failed, meta.failed());
+        assertEquals(failureStatus, meta.status());
```
[the remark this answers is not in the document]
Comment (randered) [reply hedera-node/hedera-token-service-impl/src/test/java/com/hedera/node/app/service/token/impl/test/PreHandleTokenCreateTest.java:386]: Hopefully other classes will use it as well. Moved to MetaAssertion Class.

**Review by dimitar-dinev — APPROVED** [review]
LGTM!

File: hedera-node/hedera-mono-service/src/testFixtures/java/com/hedera/test/factories/scenarios/TokenCreateScenarios.java [hunk 2]
```diff
@@ -173,6 +183,70 @@ public PlatformTxnAccessor platformTxn() throws Throwable {
...
+                            .plusCustomFee(
+                                    royaltyFee(
+                                            1,
+                                            2,
+                                            new FixedFeeSpec(1, MISSING_ENTITY_ID),
+                                            collector,
+                                            false))
+                            .get());
+        }
+    },
```
Comment (mhess-swl) [inline hedera-node/hedera-mono-service/src/testFixtures/java/com/hedera/test/factories/scenarios/TokenCreateScenarios.java:249]: @tinker-michaelj @Neeharika-Sompalli can you verify these new scenarios are implemented and tested correctly?

**Review by dimitar-dinev — APPROVED** [review]

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java [hunk 3]
```diff
@@ -39,13 +48,34 @@ public class TokenCreateHandler implements TransactionHandler {
...
      */
     public TransactionMetadata preHandle(
-            @NonNull final TransactionBody txBody, @NonNull final AccountID payer) {
-        throw new UnsupportedOperationException("Not implemented");
+            @NonNull final TransactionBody txBody,
+            @NonNull final AccountID payer,
+            @NonNull final AccountKeyLookup accountStore) {
+        final var tokenCreateTxnBody = txBody.getTokenCreation();
+        final var customFees = tokenCreateTxnBody.getCustomFeesList();
+        final var treasuryId = tokenCreateTxnBody.getTreasury();
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java:62]: Probably should check if  `tokenCreateTxnBody.hasTreasury()` before ?

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java [hunk 4]
```diff
@@ -39,13 +48,34 @@ public class TokenCreateHandler implements TransactionHandler {
...
+        final var autoRenewalAccountId = tokenCreateTxnBody.getAutoRenewAccount();
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java:63]: same here

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java [hunk 5]
```diff
@@ -39,13 +48,34 @@ public class TokenCreateHandler implements TransactionHandler {
...
+        final var meta =
+                new SigTransactionMetadataBuilder(accountStore).payerKeyFor(payer).txnBody(txBody);
+
+        meta.addNonPayerKey(treasuryId, INVALID_TREASURY_ACCOUNT_FOR_TOKEN);
+
+        if (tokenCreateTxnBody.hasAutoRenewAccount()) {
+            meta.addNonPayerKey(autoRenewalAccountId, INVALID_AUTORENEW_ACCOUNT);
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java:70]: You could add 
```
final var autoRenewalAccountId = tokenCreateTxnBody.getAutoRenewAccount();
```
inside the if condition

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java [hunk 6]
```diff
@@ -60,4 +89,62 @@ public TransactionMetadata preHandle(
...
+     * @param collector the ID of the collector
+     * @param alwaysAdd if true, will always add the key
+     */
+    private void addAccount(
+            final SigTransactionMetadataBuilder meta,
+            final AccountID collector,
+            final boolean alwaysAdd) {
+
+        if (alwaysAdd) {
+            meta.addNonPayerKey(collector);
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java:145]: I think you should also pass in `INVALID_CUSTOM_FEE_COLLECTOR` here right ?

**Review by Neeharika-Sompalli — DISMISSED** [review]
LGTM ! I have a question on the status code passing into `addNonPayerKey`

File: hedera-node/hedera-mono-service/src/testFixtures/java/com/hedera/test/factories/scenarios/TokenCreateScenarios.java (hunk 2 above)
Comment (Neeharika-Sompalli) [reply hedera-node/hedera-mono-service/src/testFixtures/java/com/hedera/test/factories/scenarios/TokenCreateScenarios.java:249]: I think they are adding more scenarios. Seems fine to me

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java (hunk 6 above)
Comment (dimitar-dinev) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenCreateHandler.java]: Oh, yes, great catch --- fixed this and added a new scenario to test it

**Review by Neeharika-Sompalli — APPROVED** [review]

**Review by mhess-swl — APPROVED** [review]

**Pull request merged**